Add plugins

vieux opened the pull request:
let's keep in small for now and only add a minimum parameters needed.

stevvooe commented on api/specs.proto:
```diff
@@ -122,6 +123,17 @@ message NetworkAttachmentSpec {
+	string image = 1;
+
+	// enabled defines if the plugin should be enabled or not.
+	bool enabled = 2;
```
Is default enabled? I think we went with Disabled in plugin API.

vieux replied:
`docker plugin install` defaults to enable. and it's also `Enabled` in `docker/docker/api/types/plugin.go`

stevvooe replied:
That is very backwards. Make sure the zero-value is the default.

anusha-ragunathan replied:
We want to enable plugins by default because:
1. After a `docker plugin install`, subsystems/containers can start using the plugin right away. Zero configuration model.
2. Every restart of docker-engine will automatically start plugins alongside, so that containers dont have ordering dependency on plugins lifecycle.

In the event that plugins are deemed to be buggy or unnecessary, they are disabled.

stevvooe replied:
If you want them enabled by default, then use a `disabled` field so the zero-value is the default.

stevvooe commented:
Let's make sure this can be implemented over the engine-api.

vieux commented:
@stevvooe I have a branch in docker/docker already, so it is implementable over the engine-api.

vieux commented:
ping @stevvooe

vieux commented:
ping @anusha-ragunathan

stevvooe commented:
@vieux Needs an engine-api (or whatever we call that now) implementation so we can test with swarmkit cli.

vieux commented:
@stevvooe: something like https://github.com/docker/docker/compare/master...vieux:swarm_plugins ?

anusha-ragunathan commented:
LGTM

vieux commented:
sorry @stevvooe I'm not sure what you want, can you tell me more ?

stevvooe commented:
@vieux Make it so it works with swarmctl/swarmd. Let's also make sure the APIs match up.

vieux commented:
@stevvooe I pushed some WIP code, can you please tell me if I'm going in the right direction ?

I added a `agent/exec/plugin` pkg with a `controller` and `adapter` for `plugin`
I also moved `agent/exec/container/executor.go` to `agent/exec/docker/executor.go` since in my opinion, the executor should be shared between containers and plugins, and it's really a `docker` executor than a `container` executor.

WDYT ?

/cc @anusha-ragunathan @tiborvass

stevvooe commented:
> the executor should be shared between containers and plugins, and it's really a docker executor than a container executor

This is a fine split.

stevvooe commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+		return err
+	}
+
+	<-pctx.Done()
```
This doesn't seem to be actually waiting on plugin state. This is almost pointless.

vieux replied:
there is no wait for `plugin` we basically want to wait forever

stevvooe replied:
What happens when the plugin fails?

tiborvass commented:
@vieux I agree the more we can make plugin == container the better.

anusha-ragunathan commented on agent/exec/docker/executor.go:
```diff
@@ -87,9 +89,21 @@ func (e *executor) Configure(ctx context.Context, node *api.Node) error {
 
 // Controller returns a docker container controller.
```
Should be a "container or plugin controller"

vieux replied:
done

anusha-ragunathan commented on agent/exec/plugin/adapter.go:
```diff
@@ -0,0 +1,48 @@
+
+func (p *pluginConfig) pluginInstallOptions() types.PluginInstallOptions {
+	return types.PluginInstallOptions{
+		Disabled:             true,
```
Plugins should be enabled by default.

vieux replied:
I agree, but the controller has 2 methods `Prepare` and `Start` I think Prepare should only install and start actually enable it don't you think ?

anusha-ragunathan replied:
In which case, I would rename `install` to `pull`

anusha-ragunathan commented on agent/exec/plugin/adapter.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+func (p *pluginAdapter) remove(ctx context.Context) error {
+	return p.client.PluginRemove(ctx, p.plugin.image(), types.PluginRemoveOptions{Force: true})
```
Why `force` remove?

vieux replied:
removed the force, you're probably right, there is no need

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (r *controller) Task() (*api.Task, error) {
+	return r.task, nil
```
no need to return error.

vieux replied:
done

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+	return nil
+}
+
+// Close the controller and clean up any ephemeral resources.
```
What "ephemeral resources" are we referring to?

vieux replied:
leftover copy paste from container controller. I removed it.

stevvooe replied:
These are resources associated with the controller itself that might be created each time the plugin is executed.

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	if err := r.adapter.enable(ctx); err != nil {
+		return errors.Wrap(err, "enabling plugin failed")
```
Why error wrapping only for enable?

stevvooe replied:
Is the error propagated up the stack?

stevvooe replied:
A little more context: we wrapped errors where we were seeing context hiding. Typically, you should wrap if you need to know where in the controller the error may have originated.

stevvooe replied:
@anusha-ragunathan What issue do you see with error wrapping?

anusha-ragunathan replied:
If we have error logprefixes such as "plugin enable"/"plugin disable" and so on at the docker level, then its useful for all clients, including swarm. That way there's good context in the error and wont require wrapping.

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+		return err
+	}
+
+	return nil
```
return `r.adapter.disable` ?

vieux replied:
right

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+// Terminate the plugin.
+func (r *controller) Terminate(ctx context.Context) error {
```
Terminate and Shutdown are doing the same thing. What's the execution path of these two?

stevvooe replied:
Shutdown should wait for graceful exit. Terminate should _terminate_.

anusha-ragunathan replied:
atm, disable does a SIGKILL. There's no graceful exit. I'll open a PR in docker/docker to address that. @vieux : Please add a TODO here so that it can be addressed accordingly, after the docker PR.

anusha-ragunathan commented on agent/exec/plugin/controller.go:
```diff
@@ -0,0 +1,148 @@
+	}, nil
+}
+
+func (r *controller) Task() (*api.Task, error) {
```
nit: `c` or `ctr` would be better than `r`.

stevvooe replied:
`c` is probably okay. `r` is leftover from when this `controller` had a different name.

anusha-ragunathan commented on agent/exec/plugin/errors.go:
```diff
@@ -0,0 +1,8 @@
+
+var (
+	// ErrImageRequired returned if a task is missing the image definition.
+	ErrImageRequired = errors.New("dockerexec: image required")
```
can move to `exec/errors.go`, so that both container and plugin can use it.

vieux replied:
will do

vieux commented on agent/exec/plugin/adapter.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+func newPluginAdapter(client engineapi.APIClient, task *api.Task) (*pluginAdapter, error) {
+	plgn, err := newPluginConfig(task)
```
I kept `plgn ` to be consistent with the existing `container` controller using `ctr`

vieux commented on agent/exec/docker/executor.go:
```diff
@@ -87,9 +89,21 @@ func (e *executor) Configure(ctx context.Context, node *api.Node) error {
+
+	if t.Spec.GetPlugin() != nil {
+		ctlr, err = plugin.NewController(e.client, t)
+		if err != nil {
```
done

anusha-ragunathan commented on Makefile:
```diff
@@ -16,6 +16,8 @@ BINARIES=$(addprefix bin/,$(COMMANDS))
 
 GO_LDFLAGS=-ldflags "-X `go list ./version`.Version=$(VERSION)"
 
+DOCKER_BUILDTAGS+="experimental"
```
Where's the tag used?

vieux replied:
swarmkit imports the `engine-api` we need this tag to have the plugins endpoints.

stevvooe replied:
@anusha-ragunathan I was more interested in that for the interim.

Do we have some way to detect if the engine supports plugin API?

anusha-ragunathan replied:
No, any API call will simply fail if engine doesnt support it. For nicer UX, we could populate the daemon's capabilities in `docker info`. This issue probably applies to all new endpoints like `checkpoint`, `system`, etc

vieux commented:
ping @aluzzardi @stevvooe @tiborvass now ready for review. unit & integrations added.

stevvooe commented:
@vieux @anusha-ragunathan Is there a docker PR for this yet? I looked at @vieux diff and it looks like we've overloaded services, which might might not be quite right.

vieux commented:
@stevvooe no PR in docker/docker yet, my initial plan was to simply update the API here and do most of the work in docker/docker, but you requested to have the changes in swarmctl as well.

The PR on docker/docker will likely do the same thing, with @al and @tiborvass .

`docker service create --runtime plugin --mode global vieux/sshfs`

I just rebased.

aluzzardi commented:
@stevvooe Wasn't that the original intent for task runtimes? (e.g. having a plugin runtime or whatnot)

stevvooe commented:
@aluzzardi At the data structure level, yes. I wasn't sure if this was what we were doing with the UX.

vieux commented:
@aluzzardi @stevvooe anything I can do to move this PR forward ?

anusha-ragunathan commented:
@stevvooe : Having docker service manage a `plugin runtime` seems like the right thing to do, rather than introducing a new object for swarm to manage.

aluzzardi commented:
@stevvooe should I take that as an LGTM?

vieux commented:
rebased 👼

stevvooe commented:
@aluzzardi There are _tons_ of options in the plugin API. I don't see how this can work with `PluginSpec` only having two fields (https://github.com/docker/swarmkit/pull/1545/files#diff-c47806c1c15110e42135ad3e8ef9dabaR133).

vieux commented:
@stevvooe the current use case we want to handle is `docker plugin install`

```
$ docker plugin install --help

Usage:	docker plugin install [OPTIONS] PLUGIN

Install a plugin

Options:
      --disable                 Do not enable the plugin on install
      --grant-all-permissions   Grant all permissions necessary to run the plugin
      --help                    Print usage
```

`--grant-all-permissions` is set to true all the time when using swarmkit so we only want to support `--disable`

I don't see what `tons of options in the plugin API.` you are refereing to.

stevvooe commented:
@vieux Plugin types are littered with `Settable` fields (https://github.com/docker/docker/blob/master/api/types/plugin_env.go#L20). How can one set them with this spec?

vieux commented:
@stevvooe `docker plugin set` it's not even implemented in docker yet, so I think we could worry about that later

stevvooe commented:
> @stevvooe docker plugin set it's not even implemented in docker yet, so I think we could worry about that later

How does one configure the plugins for use? Do they all rely on opts passed to the container?

aaronlehmann commented:
> `--grant-all-permissions` is set to true all the time when using swarmkit so we only want to support --disable

This is really scary :/. Our original design goal was for workers not to implicitly trust managers with any privileged access, but I feel we're slipping away from that.

cc @diogomonica

vieux commented:
/cc @aluzzardi

diogomonica commented:
huh, sad that this is the first time I'm pinged on this :(

@vieux would love to go over how this works, and if there is a way of defanging `grant-all-permissions`

anusha-ragunathan commented:
@aaronlehmann @diogomonica : @tiborvass is working on breaking down permissions into finer grains and disallowing them by default. So lets not panic :)

stevvooe commented:
@anusha-ragunathan So, swarmkit tasks represent an idempotent, executable action, like running a container or process. The problem I see with plugins is that `install` and `run` aren't separated. Nearly every extant runtime model I've seen separates these two steps. We do so in the `Controller` interface, with `Prepare` and `Start`, and I suspect we'll find that with plugins, separating these steps is critical, as well.

As far as plugin execution, I would expect it to work as follows:

```
name := NameOfPlugin
PluginSubsystem.Install(name)

config1 := PluginConfigOne(name)
PluginSubsystem.Start(config1)

config2 := PluginConfigTwo(name)
PluginSubsystem.Start(config2)
```

From the perspective of swarmkit, what we are calling the `PluginSpec` would have to represent the runtime config of the plugin. This model will also allow multiple instances of a plugin with different configurations.

My point here is that the plugins API data structures have a bunch of stuff we aren't using (`docker plugin set`) and we'll have to eventually represent in the swarmkit task runtime. We either need to fix the data structures involved in plugins to remove this unused content or consider these implications in the design of the task runtime model.

anusha-ragunathan commented:
Plugins *do* have the install and runtime separated. The default behaviour in docker/docker is to combine the two. There's an option to simply install, without running (enabling) the plugin.

The plugin execution workflow you've described is possible (just that `docker plugin set` is not implemented in docker yet). If a plugin can be configured in 2 different ways by a user, then both instances will be allowed to run at a time. 

If the main issue is that, this PR cannot progress since `docker plugin set` is not implemented yet, then that's okay. We can keep this open and resume work when the docker/docker PR for `set` is completed.

@stevvooe : Does that work for you?

aaronlehmann commented:
@vieux: Should this be closed now that we're going with the approach in #2014?

aaronlehmann commented:
ping @vieux

aaronlehmann commented:
Going once, going twice.
